File: java/leetcode/unique-paths-ii.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m = obstacleGrid.length;
        int n = obstacleGrid.length > 0 ? obstacleGrid[0].length : 0;
        int[][] paths = new int[m][n];

        // If first coordinate is an obstacle this is unsolvable
        paths[0][0] = obstacleGrid[0][0] == 1 ? 0 : 1;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (obstacleGrid[i][j] == 1) {
                    continue;
                }
                if (i > 0) {
                    paths[i][j] += paths[i - 1][j];
                }
                if (j > 0) {
                    paths[i][j] += paths[i][j - 1];
                }
            }
        }
        return paths[m - 1][n - 1];
    }
}

File: python/leetcode/unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        m = len(obstacleGrid)
        n = len(obstacleGrid[0]) if len(obstacleGrid) > 0 else 0
        paths = []
        for i in range(m):
            paths.append([0] * n)
        paths[0][0] = 1 if obstacleGrid[0][0] == 0 else 0

        for i in range(m):
            for j in range(n):
                if obstacleGrid[i][j] == 1:
                    continue
                if i > 0:
                    paths[i][j] += paths[i - 1][j]
                if j > 0:
                    paths[i][j] += paths[i][j - 1]
        print(paths)
        return paths[m - 1][n - 1]
